update: wpb-22590 planning page for Wire server deploy and update images (#99)
wireapp/wire-docs: src/how-to/install/planning.md

# Implementation plan

There are three types of installation targets for Wire: **WIAB Dev (formerly WIAB Demo)**, **WIAB Staging**, and **Production**. This page helps you choose the right one and points you to the corresponding guides.

## Deployment options at a glance

| Option          | Purpose                                  | Scale & Topology                          | Persistence           | Typical Audience                     |
|-----------------|-------------------------------------------|-------------------------------------------|------------------------|--------------------------------------|
| WIAB Dev        | Quick functional demo on a single VM     | 1 Ubuntu 24.04 VM with Minikube           | No (ephemeral data)    | Developers, PoC, quick evaluations   |
| WIAB Staging    | Realistic multi-VM, non-production setup | ~7 Ubuntu 22.04 VMs on 1 physical host    | Yes (multi-service)    | Infra/ops testing, training     |

> | Production      | High-availability, long-term deployment  | Separate DB VMs + 2 Kubernetes cluster      | Yes (HA, backups, etc.)| Production operations                |

Reviewer: On production there is no indicator that says only ubuntu 22.04 version is supported

Reviewer: I would have a different header,  `purpose` and `Typical Audience` pretty much the same thing.  `Scale & Topology` mixes two concerns (hardware footprint + k8s setup) into one terse cell.

Options, Best for, Min. hardware, VMs, Persistent Data, Production-Grade.

As an example of  Option: WIAB Dev, Best for: Demos, feature evaluation, dev testing, Min. hardware:  1 VM (Ubuntu 24.04, ≥16 CPU/16 GB RAM) , VMs: 1, Persistent Data: No - ephemeral only, Production-Grade: No

> | Production      | High-availability, long-term deployment  | Separate DB VMs + 2 Kubernetes cluster      | Yes (HA, backups, etc.)| Production operations                |

Reviewer 2: cluster might need to be plural / culster**s** ?

- **WIAB Dev** is a self-contained, single-node demo environment (previously called *WIAB Demo*). All core services run on one VM, using in-cluster, ephemeral datastores. It is ideal for trying out functionality but **not suitable for production**.
- **WIAB Staging** reproduces the production architecture using a fixed set of virtual machines hosted on a single physical box. It is persistent but still a single failure domain and therefore **not production-grade**.
- **Production** uses dedicated virtual machines for databases and a Kubernetes cluster for Wire services. It is designed for **durability, scale, and high availability**.

> **Important:** There is no supported migration path from WIAB Dev or WIAB Staging to Production. Treat them as disposable environments.

## WIAB Dev (single-VM Wire-in-a-Box)

**Use this if:**

- You want to see Wire running quickly.
- You are evaluating features or doing development integration.

> - You are happy to lose the data when you tear down the VM or during restart

Reviewer 2: ```suggestion
- You are happy to lose the data when you tear down the VM or during restart.
```

> **Internet requirement for WIAB Dev:** The WIAB Dev playbook assumes that the deploy_node has outbound internet connectivity during installation. Tooling such as Minikube, Docker, kubectl, Python packages, and system packages are downloaded from public repositories and are **not shipped inside the artifact bundle**. After installation, you can lock down outbound access more aggressively, but Let’s Encrypt certificate renewal (if used) will still require periodic internet access.

## WIAB Staging (multi-VM KVM-based test cluster)

**Use this if:**

- You want to rehearse a production-style deployment or upgrade process.
- You want realistic separation of roles (kubernetes vs data services vs assets) on multiple VMs.
- You want a staging Wire server with stateful services.

**Key characteristics:**

- One physical machine with hardware virtualization support (KVM).
- Typically **7 VMs** (Ubuntu 22.04) to simulate a production layout:
  - `assethost`  – asset and artifact host.
  - `kubenode1–3`  – Kubernetes control-plane and worker nodes running Wire services and calling.

>   - `datanode1–3`  – data services such as Cassandra, RabbitMQ, Postgresql, Minio and Elasticsearch.

Reviewer 2: Postgre**SQL** and Min**IO** for consistent capitalization.

![Wire in a Box Staging Architecture](img/architecture-wiab-stag.png)

> **Internet requirement for WIAB Staging provisioning:** If you use the WIAB Staging provisioning Ansible playbook to create the VMs on your physical host, that provisioning step will require internet access to download the supporting tools and images it needs (for example VM cloud images and KVM/libvirt tooling). Once the VMs and the Wire artifact are in place, subsequent deployment stages can run with limited or no internet, depending on your TLS/certificate setup.

For detailed requirements, VM sizing, playbook usage, and follow-up cluster deployment steps, see:

- **Guide:** [`WIAB Staging Deployment`](wiab-staging.md)

## Production installation (persistent data, high-availability)

What you need:

- a way to create **DNS records** for your domain name (e.g. `wire.example.com`)
- a way to create **SSL/TLS certificates** for your domain name (to allow connecting via `https://wire.example.com`)
- two **kubernetes clusters with at least 3 worker nodes each** (some cloud providers offer a managed kubernetes cluster these days)

>   - First i.e. primary kubernetes cluster will be used by wire-server and supporting services. Second kubernetes cluster aka secondary would be used for calling workload. Calling infrastructure is designed to operate in a Zero trust / DMZ environment, so that the networking requirements of calling can be separated from the sensitive data stored in your Wire backend i.e. primary k8s cluster. Read more about Wire architecture [here](../../understand/overview.md#backend-routing)

Reviewer 2: ```suggestion
  - First (primary) kubernetes cluster will be used by wire-server and supporting services. Second kubernetes cluster aka secondary would be used for calling workload. Calling infrastructure is designed to operate in a Zero trust / DMZ environment, so that the networking requirements of calling can be separated from the sensitive data stored in your Wire backend (primary) k8s cluster. Read more about Wire architecture [here](../../understand/overview.md#backend-routing)
```

Microscopic syntax nitpick.

> - minimum **23 virtual machines** for components outside kubernetes (cassandra, minio, elasticsearch, rabbitmq, postgresql)

Reviewer 2: This says these 23 VMs are "for components outside kubernetes," listing only database/storage services. 

But the actual 23 count from the table includes:

* 15 database/storage VMs (Cassandra 3 + PostgreSQL 3 + MinIO 3 + ES 3 + RabbitMQ 3)
* 6 Kubernetes VMs (Primary 3 + Secondary 3)
* 2 infrastructure VMs (Admin Host 1 + Asset Host 1)

So only 15 of the 23 are "outside kubernetes." 
The Kubernetes nodes and Admin/Asset hosts account for the other 8.

Very possible I'm missing something, just thought I'd point it out.

|------------------------------------------------------|----------|--------------|---------------|-------------------|
| Cassandra                                            | 3        | 1            | 2             | 80                |
| Postgresql                                           | 3        | 1            | 2             | 80                |
| MinIO                                                | 3        | 1            | 2             | 400               |
| ElasticSearch                                        | 3        | 1            | 2             | 60                |
| RabbitMQ                                             | 3        | 2            | 4             | 80                |
| Primary Kubernetes (k8s control/worker node)         | 3        | 6            | 8             | 80                |
| Admin Host                                           | 1        | 1            | 4             | 90                |
| Asset Host                                           | 1        | 1            | 4             | 100               |

> | Stateful Services Totals per physical node           | -        | 14 CPU Cores | 32 GB Memory  | 970 GB Disk Space |

Reviewer 2: Totals don't add up for Memory (certain)  I think (CPU cores is correct)

Reviewer 2: RAM should be 28G

> | Secondary Kubernetes (Calling Services)**            | 3        | 4            | 10            | 70                |

Reviewer: ```suggestion
| Secondary Kubernetes (Calling Services)           | 3        | 4            | 10            | 70                |
```

> | Single Server Totals                                 |  -       | 18 CPU Cores | 44 GB Memory  | 1060 GB Disk Space |

Reviewer 2: 970 + 70 = 1040, not 1060.

> Notes:

> > - Secondary kubernetes hosts may need more resources to support heavy conference calling. For concurrent SFT users (SFT = Selective Forwarding Turn-server, ie. Conference calling), we recommend an extra 3% of CPU allocation, evenly distributed across the nodes (i.e. 1% more CPU per kubernetes server). So for every 100 users plan on adding one CPU core on each Kubernetes node. The SFT component runs inside of Kubernetes, and does not require a separate virtual machine for operation.  

Reviewer 2: ```suggestion
> - Secondary kubernetes hosts may need more resources to support heavy conference calling. For concurrent SFT users (SFT = «Selective Forwarding Turn» server, ie. Conference calling), we recommend an extra 3% of CPU allocation, evenly distributed across the nodes (i.e. 1% more CPU per kubernetes server). So for every 100 users plan on adding one CPU core on each Kubernetes node. The SFT component runs inside of Kubernetes, and does not require a separate virtual machine for operation.  
```

Separating `turn` from `server` because «turn server» is a specific thing that's different from SFT in Wire context.

> > - Admin Host and Asset Host can run on any one of the 3 servers, but the respective server must allocate additional resources as indicated in the table above.  

Reviewer: Which three servers? And one must need to allocate more resources only if they decided to those three servers(?) to configure admin and asset host?

Author: These 3 servers are referring to 3 physical machines mentioned above. One should consider the resources mentioned above which counts for adminhost and assethost.

> > - Wire requires the usage of CPUs built on the AMD64 architecture, and assumes these are running at the equivalent CPU/Memory performance of a recent Intel/AMD server system clocked at at least 2.5Ghz. For reference, please refer to https://cpubenchmark.net/high_end_cpus.html  

Reviewer: Is the cpubenchmark website standard?

Author: Its a resource we have been using internally for quite sometime now

> > - Wire requires the usage of CPUs built on the AMD64 architecture, and assumes these are running at the equivalent CPU/Memory performance of a recent Intel/AMD server system clocked at at least 2.5Ghz. For reference, please refer to https://cpubenchmark.net/high_end_cpus.html  

Reviewer 2: ```suggestion
> - Wire requires the usage of CPUs built on the AMD64 architecture, and assumes these are running at the equivalent CPU/Memory performance of a recent Intel/AMD server system clocked at at least 2.5 Ghz. For reference, please refer to https://cpubenchmark.net/high_end_cpus.html  
```

Microscopic syntax nitpick.

> > - Systems which are running under virtualization must be using Hardware Virtualization Support.  

Reviewer: Too much of` virtualization` in one sentence

> > - Each Virtual core should correspond to a reservation of one physical core, not a thread.  

Reviewer: So the assumptions is reader knows the difference between physical core and a thread?

Author: Since this is required for production setup and the basic requirement is to have 3 physical machines. It should give an idea when a user would consider hardware. How would you like to phrase it?

> > - Note that some performance will be lost with mitigations for Spectre, Meltdown, and other security patching, so plan reserve capacity as appropriate.  

Reviewer: What are the signals to measure the capacity?

> > - Disk space requirements are referring to formatted, local disk space. The use of network file storage solutions may be incompatible with Wire deployment and operation. Let us know if you are considering this type of solution.  

Reviewer: What kind of storage solutions are not compatible here?

Author: This is a general recommendation, we have not tested all network file storage solutions. It can depend on the solution from a user.

> > - Disk space must be SSD or NVMe. We recommend against using spinning disks. 

Reviewer: if it must be SSD or NVMe then why not just saying spinning disk is not supported rather using not recommended
